helix-view/src/document.rs

```diff
@@ -140,46 +140,66 @@ use helix_lsp::lsp;
 use url::Url;
 
 impl Document {
-    pub fn new(text: Rope) -> Self {
-        let changes = ChangeSet::new(&text);
-        let old_state = None;
+    pub fn new() -> Self {
+        let text = Rope::from("\n");
 
-        Self {
-            id: DocumentId::default(),
-            path: None,
-            text,
-            selections: HashMap::default(),
-            mode: Mode::Normal,
-            restore_cursor: false,
-            syntax: None,
-            language: None,
-            changes,
-            old_state,
-            diagnostics: Vec::new(),
-            version: 0,
-            history: Cell::new(History::default()),
-            last_saved_revision: 0,
-            language_server: None,
-        }
+        Self::from(text)
     }
 
     // TODO: async fn?
-    pub fn load(path: PathBuf) -> Result<Self, Error> {
-        use std::{fs::File, io::BufReader};
+    pub fn open(path: PathBuf) -> Result<Self, Error> {
+        use std::{
+            fs::File,
+            io::{BufReader, Read},
+        };
 
-        let doc = if !path.exists() {
-            Rope::from("\n")
-        } else {
-            let file = File::open(&path).context(format!("unable to open {:?}", path))?;
-            let mut doc = Rope::from_reader(BufReader::new(file))?;
-            // add missing newline at the end of file
-            if doc.len_bytes() == 0 || doc.byte(doc.len_bytes() - 1) != b'\n' {
-                doc.insert_char(doc.len_chars(), '\n');
+        if !path.exists() {
+            return Ok(Self::new());
+        }
+
+        let mut buf = vec![0u8; 1 << 14];
+        let mut buf_str = String::with_capacity(1 << 14);
+        let mut builder = RopeBuilder::new();
+        let mut file = BufReader::new(File::open(&path)?);
+        let mut encoding = {
+            if let Ok(n) = file.read(&mut buf) {
+                let mut encoding_detector = chardetng::EncodingDetector::new();
+                encoding_detector.feed(&buf, false);
+
+                let encoding = encoding_detector.guess(None, true);
+
+                let (temp, ..) = encoding.decode(&buf);
+                buf = temp.as_bytes().to_vec();
+                buf_str.clear();
+                for &byte in &buf[..n] {
+                    buf_str.push(byte as char);
+                }
+                builder.append(&buf_str);
+
+                Some(encoding)
+            } else {
+                None
             }
-            doc
         };
 
-        let mut doc = Self::new(doc);
+        while let Ok(n) = file.read(&mut buf) {
+            if n == 0 {
+                break;
+            }
+
+            let (temp, ..) = encoding.unwrap().decode(&buf);
+            buf = temp.as_bytes().to_vec();
```

Comment:
I don't fully get this, you get back a Cow<str> but instead of using it in the rope builder you convert to bytes then back to str. The decode result is directly usable in the builder, you just need to deref it

Reply:
Honestly, I rushed this code in just a few minutes. 😅 I was eager to get it working rather than checking if the implementation was good, but I still might have missed that anyway so thanks!